### DISCUSSION TITLE
Text content did not match

### DISCUSSION THREAD

**Discussion by binajmen** [sergiodxa/remix-i18next#74] [Q&A] [answered]
Hi @sergiodxa,

I tried your library, thinking about including i18n in my project before it is too late. But I'm already facing a weird issue:
```ts
Warning: Text content did not match. Server: "Bonjour" Client: "Hello"
```
I understand there is a clash during the hydration, with a missmatch between the server and the client. But I don't get why.. It seems I followed everything from scratch.

A sandbox repo is available here: https://github.com/binajmen/remix-i18n-sandbox/tree/main/app/routes/

FYI, I'm in Belgium, so perhaps there is a clash between `/en` in the path, my geoloc, and my OS language (english)?

A secondary question would be: how do you handle locale selection when using `lng` in path instead of the `searchParams`?

I noticed this code snippet in another discussion, but it is unclear how I could make is work with a path instead.
[remix-i18next/src/i18next.ts](https://github.com/sergiodxa/remix-i18next/blob/8ab6361122147ea0d63c88ef4fb98295bd8473a9/src/i18next.ts#L52-L61)

Thank you for your help 🙏

**Comment by sergiodxa** [comment, marked as answer]
This happens because you didn't loaded the same translations on the server and the client, ensure since you're using a `$lang` route to detect the user language, you should not detect it using `i18n.getLocale` because that will not respect the URL, it will read it from your search params, session, cookie or headers.

You either use i18n.getLocale or use the params on the URL.

**Reply by binajmen** [reply]
So you're suggesting the following?
```ts
export let loader: LoaderFunction = async ({ request, params }) => {
  const { lang } = params;
  invariant(typeof lang === "string");

  // let locale = await i18next.getLocale(request);

  return json<LoaderData>({ locale: lang });
};

export default function App() {
  let { locale } = useLoaderData<LoaderData>();
  let { i18n } = useTranslation();

  useChangeLanguage(locale);

  return (
    <html lang={locale} dir={i18n.dir()}>
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```

**Reply by binajmen** [reply]
Oh in fact I must modify `entry.server.tsx` as well:
```ts
  // let lng = await i18next.getLocale(request);
  let lng = remixContext.matches[0].params.lang!;
```
It seems a bit hacky to retrieve the lang that way, but I suppose this is the only way, correct?

Would appreciate to get your feedback on that strategy. Thank you for your help, really appreciated 🙏

**Reply by sergiodxa** [reply]
Yes, something like that should work!

**Reply by binajmen** [reply]
Yes I ended up with:
```ts
const lng = remixContext.matches[0].params.lang ?? i18n.fallbackLng;
  ```
